Make rebuilding packages smarter

olavphibm opened the pull request:
This changes the default behavior of build-packages command to only build packages when needed.
Any of the following criteria is enough to rebuild a package:
- It has no previously cached build results
- Any of Its source files (yaml, spec, etc...) were modified
- It has build dependencies that were rebuilt
- The option --force-rebuild was used (forces rebuild of all packages at once)

Commit:
Archive build artifacts to a package specific directory

This makes it straightforward to know which package generated which
RPMs. It also needed to know whether a package already has results from
previous builds. Before, we had no permanent information to link an RPM
to a package name after the Package object was destroyed.

olavphibm commented:
The first two commits belong to https://github.com/open-power-host-os/builds/pull/175

alanoe commented on lib/package.py:
```diff
@@ -56,7 +56,8 @@ def __init__(self, name):
         self.install_dependencies = []
         self.build_dependencies = []
-        self.result_packages = []
+        self.artifacts_dir = os.path.join(
```
pls be consistent in the terminology, use either "artifacts" or "results". I prefer the latter, as it is a more universally know term. "artifacts" is Jenkins terminology.

olavphibm replied:
The class' name is `Package`, so `Package.results` does not make much sense to me. I admit `Package.artifacts` doesn't either. What do you think of `Package.build_results_dir`?

alanoe replied:
+1

alanoe commented on config.yaml:
```diff
@@ -21,6 +21,7 @@ default:
    --plugin-option=tmpfs:required_ram_mb=39800"
  commit_updates: True
  push_updates: True
+ build_dependencies: True
```
I don't think that this global attribute is needed, I'd prefer to have a clean_build_dependencies_cache boolean option instead. If this option is disabled, the dependencies reuse would be dynamic instead of the static approach you suggest here. We would not assume that all dependencies are already pre-built or not; for each package, the build system auto-detects if the dependencies of a package are already built in the system and reuses them if they are. To achieve this, we just need to choose an adequate primary key for a package build (eg `<package name`>-`<commit ID`>) and somehow keep an updated table of the built packages. One way to achieve this is by saving each build package in a result dir of format <primary_key>, later parsing this directory structure to infer if the dependency package was already build or not and reuse the package if it was already built.

olavphibm replied:
I don't think using the commit ID as a primary key is a good idea. There are many other factors that may require rebuilding a package. Altering the spec file or modifying the build scripts, for example. I agree with checking if the dependencies are already built by checking if a directory `<package_name>` exists, but I think the default should be to always rebuild the dependencies and the decision not to do this should be mostly manual.

alanoe replied:
My main point here is automating the act of inferring if *each* package must be rebuilt instead of having a global manual option to do it, the commit ID was just the first idea that came in my mind. Better ideas are welcome. =)   Another (better) way is to keep the date of the last modification in the package directory and rebuild if there were any modifications since then. This would cover adding of files to SOURCES, changes in the spec and changes in the yaml. The only case this approach would not cover is a modification to the builds git repo, but then you could explicitly use the option to clean the cache I've mentioned before.

alanoe commented on lib/build_manager.py:
```diff
@@ -46,5 +48,11 @@ def __call__(self):
+            packages_order = scheduler(packages_order)
+        else:
+            LOG.info("Not scheduling build of dependencies")
+            LOG.info("Building packages: {}".format(packages_order))
```
I don't agree with this. The build scheduler should always be used, there is no reason to let user specify the build order.

olavphibm replied:
Flexibility is always nice, but yeah, I don't see any good reason to do this either.

alanoe commented on lib/mockbuilder.py:
```diff
@@ -17,6 +17,7 @@
 
```
It seems to me that, with this approach, if you desire to build gcc and 2 packages which depend on gcc, gcc will be built 3 times. This does not make sense to me. Maybe I misunderstood? 

I don't think we need to keep the built build dependencies in a package-specific directory, I'd prefer to have a directory for each package built (named after a primary key) and keep them all under a single directory, as mentioned in the other patch of this PR.

olavphibm replied:
The way I've done it, only the packages explicitly passed would be built, so passing `gcc, kernel` would only build gcc once. I'm applying your other suggestions though, so this will be redone.

I don't get your second point, doesn't your suggestion of using a primary key lead to package (and commit) specific directories?

alanoe replied:
OH> I don't get your second point, doesn't your suggestion of using a primary key lead to package (and commit) specific directories?

I was under the impression in my first quick read of the code that, if you are building A and B and C are dependencies, you'd put B and C rpms inside A dir. In that case, if C was a dependency for D, you would not find C inside D dir and rebuild C.

Commit:
Copy build results to a cache directory

This will allow reusing the results from previous builds without getting
any depending on information from the results directory.

Create property cached_build_results in Package class

Commit:
Do not rebuild packages with previous build results

By default, only packages with missing results will be rebuilt.
This patch adds an option to force rebuilding packages.

Commit:
Organize build_manager imports

Commit:
Clean cached build results before building

This makes sure build results are not mixed with previous builds
results, for example, when the spec file changes adding/removing an RPM.

Commit:
Do not rebuild packages with no modifications

Unless --force-rebuild is passed, packages with source and metadata
files older than the build results will not be rebuilt.

farosas commented on lib/rpm_package.py:
```diff
@@ -187,19 +187,25 @@ def _load(self):
             self.download_build_files = files.get('download_build_files', [])
 
+
+            rebuild_dependencies = CONF.get('default').get('force_rebuild')
```
Why can't you use `self.rebuild` here?

olavphibm replied:
This option had a slightly different meaning in a previous version of the PR.
Changing to `self.rebuild`

farosas commented on lib/packages_manager.py:
```diff
@@ -25,7 +25,14 @@
+            packages_names: names of the packages that will be processed.
+                Leave empty to discover all available packages.
+        """
+        self.packages_names = packages_names or config.discover_packages()
```
I don't get how moving the package discovery to here accomplishes what you describe in the commit message. Am I missing something here?

olavphibm replied:
This was needed for a previous version of the PR that had a slightly different functionality. I am dropping this patch.

Commit:
Refactor Mock._save_rpm method

This makes the method more generic to be used to copy the RPMs from and
to other directories, which will be done in a future patch.
- Rename it to _copy_rpms.
- Replace parameter package with source_dir and target_dir.
- Rename some variables.
- Add doc string.
- Improve logging.

Commit:
Create build results directory with default permissions

There's no reason to have loose permissions for this directory.

Commit:
Remove duplicated log message

Commit:
Separate build results copying from build method

This will allow us to copy previously cached build results without
rebuilding the package.

Commit:
Rebuild package if its build dependencies were rebuilt

alanoe commented on lib/mockbuilder.py:
```diff
@@ -17,6 +17,7 @@
 
 import datetime
```
in the commit longer msg, form => from

alanoe commented on lib/mockbuilder.py:
```diff
@@ -128,9 +128,7 @@ def _install_external_dependencies(self, package):
-                dep_rpms_glob = os.path.join(dep.build_results_dir, "*.rpm")
-                rpm_files = glob.iglob(dep_rpms_glob)
-                install = " ".join([install, " ".join(rpm_files)])
+                install = " ".join([install, " ".join(dep.cached_build_results)])
```
I'd prefer "cached_built_rpm_files" to make clearer this variable is a list of files paths.

olavphibm replied:
This is a property of the `Package` class, so we can't have RPM in its name.

alanoe replied:
The point is that "results" are a generic term, it could be logs, built packages or extra outputs (eg a yum repository config file). 

You may use cached_built_files or cached_built_files_paths.

olavphibm replied:
I've taken some time to rewrite all occurrences of "results" to "built_files" and I don't think its an improvement. It makes logs and docs harder to read. "Results", although being generic, is a simpler term.
For example: "Build results found" vs. "Previously built files found".
It is not immediately clear that "built files" resulted from a previous build, so we'd have to take extra care to explain this every time.
I'm keeping this as is.

alanoe commented on lib/package.py:
```diff
@@ -56,6 +59,7 @@ def __init__(self, name):
         self.download_source = None
         self.install_dependencies = []
         self.build_dependencies = []
+        self.build_cache_dir = os.path.join(BUILD_CACHE_DIR, self.name)
```
Not sure if we need to keep build_cache_dir in Package if it easily inferred from cache dir and package name, but up to you. The same applies to build_results_dir BTW

olavphibm replied:
@gutoyr will work on changing the cache and results paths. Let's see if it's still needed after his changes.

alanoe commented on lib/mockbuilder.py:
```diff
@@ -145,18 +145,23 @@ def _create_build_directory(self, package):
+        for rpm_file in os.listdir(source_dir):
+            if rpm_file.endswith(".rpm") and not rpm_file.endswith(".src.rpm"):
+                LOG.info("Copying RPM file: %s" % rpm_file)
+                orig = os.path.join(source_dir, rpm_file)
```
- I'd prefer to be consistent with source_dir and target_dir in these orig/dest variables names
- Also, I'd prefer to have some consistency in the names of variables which refer to the file name and to the file path. You are using rpm_file for file name and orig/dest. Either use rpm_file_name and rpm_file_path, source_file_name and source_file_path or something similar.

alanoe commented on lib/mockbuilder.py:
```diff
@@ -145,18 +145,23 @@ def _create_build_directory(self, package):
+                LOG.info("Copying RPM file: %s" % rpm_file)
+                orig = os.path.join(source_dir, rpm_file)
+                dest = os.path.join(target_dir, rpm_file)
+                shutil.copy(orig, dest)
```
It is not clear to me why you are copying instead of moving. Could you explain better? 

Minor: I'd prefer to move this functional change to a separate patch and leave only refactoring here.

alanoe commented on config.yaml:
```diff
@@ -21,6 +21,7 @@ default:
    --plugin-option=tmpfs:required_ram_mb=39800"
```
minor: In commit longer msg, imho "This patch adds an option" is clearer than "There's an option"

alanoe reviewed (changes requested):
Minor changes only

Commit:
Copy files instead of moving in Mock._save_rpm method

We have an option --keep-builddir that keeps the build directory after
a successful package build. It might be confusing for the user if we
remove some build artifacts from this directory when the build finishes.

Also, this will allow us to reuse this method when we create a build
results cache.
